agent/agentserver/server_test.go

```diff
@@ -276,13 +277,113 @@ func TestReadinessCheckHandler(t *testing.T) {
 
+func TestReadinessCheckHandlerCache(t *testing.T) {
+	testErr := errors.New("test-err")
+	type call struct {
+		success bool
+	}
+
+	for _, tc := range []struct {
+		desc              string
+		readinessCacheTTL time.Duration
+		waitInvalidation  bool
+		expCalls          []call
+		setupMocks        func(m *serverMocks)
+	}{
+		{
+			desc:              "call 1 succeeds, so second call succeeds without checks",
+			readinessCacheTTL: 10 * time.Minute,
+			waitInvalidation:  false,
+			expCalls:          []call{{success: true}, {success: true}},
+			setupMocks: func(m *serverMocks) {
+				m.sched.EXPECT().Probe().Return(nil).Times(1)
+				m.tags.EXPECT().CheckReadiness().Return(nil).Times(1)
+				m.ac.EXPECT().CheckReadiness().Return(nil).Times(1)
+			},
+		},
+		{
+			desc:              "call 1 fails, so second call performs checks",
+			readinessCacheTTL: 10 * time.Minute,
+			waitInvalidation:  false,
+			expCalls:          []call{{success: false}, {success: true}},
+			setupMocks: func(m *serverMocks) {
+				m.sched.EXPECT().Probe().Return(testErr).Times(1)
+				m.tags.EXPECT().CheckReadiness().Return(testErr).Times(1)
+				m.ac.EXPECT().CheckReadiness().Return(testErr).Times(1)
+
+				m.sched.EXPECT().Probe().Return(nil).Times(1)
+				m.tags.EXPECT().CheckReadiness().Return(nil).Times(1)
+				m.ac.EXPECT().CheckReadiness().Return(nil).Times(1)
+			},
+		},
+		{
+			desc:              "call 1 succeeds, but cache becomes invalid, so second call performs checks",
+			readinessCacheTTL: 10 * time.Minute,
+			waitInvalidation:  true,
+			expCalls:          []call{{success: true}, {success: false}},
+			setupMocks: func(m *serverMocks) {
+				m.sched.EXPECT().Probe().Return(nil).Times(1)
+				m.tags.EXPECT().CheckReadiness().Return(nil).Times(1)
+				m.ac.EXPECT().CheckReadiness().Return(nil).Times(1)
+
+				m.sched.EXPECT().Probe().Return(testErr).Times(1)
+				m.tags.EXPECT().CheckReadiness().Return(testErr).Times(1)
+				m.ac.EXPECT().CheckReadiness().Return(testErr).Times(1)
+			},
+		},
+		{
+			desc:              "call 1 succeeds, but caching is disabled, so second call performs checks",
+			readinessCacheTTL: 0,
+			waitInvalidation:  false,
+			expCalls:          []call{{success: true}, {success: false}},
+			setupMocks: func(m *serverMocks) {
+				m.sched.EXPECT().Probe().Return(nil).Times(1)
+				m.tags.EXPECT().CheckReadiness().Return(nil).Times(1)
+				m.ac.EXPECT().CheckReadiness().Return(nil).Times(1)
+
+				m.sched.EXPECT().Probe().Return(testErr).Times(1)
+				m.tags.EXPECT().CheckReadiness().Return(testErr).Times(1)
+				m.ac.EXPECT().CheckReadiness().Return(testErr).Times(1)
+			},
+		},
+	} {
+		t.Run(tc.desc, func(t *testing.T) {
+			require := require.New(t)
+			mocks, cleanup := newServerMocks(t)
+			defer cleanup()
+
+			tc.setupMocks(mocks)
+
+			s, addr := mocks.startServer(Config{readinessCacheTTL: tc.readinessCacheTTL})
+			_, err := httputil.Get(fmt.Sprintf("http://%s/readiness", addr))
+			if tc.expCalls[0].success {
```

Comment:
If it's just two hardcoded calls and you're not iterating over the slice I think it would be simpler if you just had `firstCall`, `secondCall` like you did before.

The slice of calls is nice when you test an arbitrary amount of calls as you loop over them. If the clock was mocked out you could write a test with say 3 calls where 2 are cached and the last one is refetched. Since we have to stick in that `tc.waitInvalidation` in the middle we can't loop over.

Or we could add another parameter to the call - something like `invalidateAfter` and then we could invalidate the cache after the n'th call in the call for loop.

It's not a huge deal so both options are fine with me.

Reply:
I agree. In this case we're not testing overly complex caching logic, so an arbitrary amount of calls is unnecessary. I'm reverting to using two variables instead of a slice.